packages/squiggle-lang/src/cli/commands/run.tsx

```diff
@@ -0,0 +1,290 @@
+import path from "path";
+import { FC } from "react";
+
+import { defaultEnv, Env } from "../../dists/env.js";
+import { SqLinker } from "../../public/SqLinker.js";
+import { SqProject } from "../../public/SqProject/index.js";
+import {
+  ModuleData,
+  ProjectState,
+} from "../../public/SqProject/ProjectState.js";
+import { SqModule } from "../../public/SqProject/SqModule.js";
+import { allRunnerNames, runnerByName } from "../../runners/index.js";
+import { CliPrinter } from "../CliPrinter.js";
+import { bold, red } from "../colors.js";
+import { loadSrc, myParseInt } from "../utils.js";
+
+type OutputMode = "NONE" | "RESULT_OR_BINDINGS" | "RESULT_AND_BINDINGS";
+
+type RunArgs = {
+  src: string;
+  filename?: string;
+  output: OutputMode;
+  measure?: boolean;
+  sampleCount?: string | number;
+  seed?: string;
+  profile?: boolean;
+  runner?: Parameters<typeof runnerByName>[0];
+  runnerThreads?: number;
+  showProjectState?: boolean;
+  logProjectActions?: boolean;
+};
+
+const EVAL_SOURCE_ID = "[eval]";
+
+function getLinker(): SqLinker {
+  return {
+    resolve(name, fromId) {
+      if (!name.startsWith("./") && !name.startsWith("../")) {
+        throw new Error("Only relative paths in imports are allowed");
+      }
+      const dir =
+        fromId === EVAL_SOURCE_ID ? process.cwd() : path.dirname(fromId);
+      return path.relative(process.cwd(), path.resolve(dir, name));
+    },
+    async loadModule(sourceId, hash) {
+      if (hash) {
+        throw new Error("Hashes are not supported");
+      }
+      const code = await fs.readFile(sourceId, "utf-8");
+      return new SqModule({
+        name: sourceId,
+        code,
+      });
+    },
+  };
+}
+
+const ModuleInfo: FC<{
+  state: ProjectState;
+  module: SqModule;
+  environment: Env;
+}> = ({ state, module, environment }) => {
+  const output = state.getOutput(module, environment);
+
+  return (
+    <Box gap={2}>
+      <Text>{module.name}</Text>
+      <Text>
+        {output
+          ? "✅"
+          : module.getImportModules({ state }).type === "loading"
+            ? "⌛ Loading imports"
+            : module.getImportOutputs({ state, environment }).type === "loading"
+              ? "🔄 Waiting for import outputs"
+              : "▶️ Running"}
+      </Text>
+    </Box>
+  );
+};
+
+const ModuleDataInfo: FC<{
+  state: ProjectState;
+  hash: string;
+  moduleData: ModuleData;
+  environment: Env;
+}> = ({ state, hash, moduleData, environment }) => {
+  return (
+    <Box gap={2}>
+      <Text>
+        {moduleData.type === "loaded"
+          ? "✅"
+          : moduleData.type === "loading"
+            ? "⌛"
+            : "❌"}
+      </Text>
+      {moduleData.type === "loaded" ? (
+        <ModuleInfo
+          module={moduleData.value}
+          state={state}
+          environment={environment}
+        />
+      ) : (
+        <Text>{hash}</Text>
+      )}
+    </Box>
+  );
+};
+
+const StateGraph: FC<{ state: ProjectState; environment: Env }> = ({
```

Comment:
React in CLI!

Works surprisingly well, e.g. redraws correctly even if something else was printed between renders.

Bundle size is a bit heavy, but I only use it in CLI files, so it shouldn't affect components bundle size.

Reply:
Huh, that's neat. I look forward to using it. Surprised it wasn't harder.